File: src/frontend/frontend_service/src/grpc/query/BillingClient.js
import * as grpcWeb from 'grpc-web';
import { FieldMask } from 'google-protobuf/google/protobuf/field_mask_pb';
import { Empty } from 'google-protobuf/google/protobuf/empty_pb';

import {
  BillingClient, 

  BillingData,
  BillingDataList,
  BillingsRequest
} from '../../proto/query_webclient_grpc_web_pb';

String.prototype.capitalize = function() {
  return this.charAt(0).toUpperCase() + this.slice(1);
}

const BillingService = new BillingClient(
                          'https://' + window.location.hostname + ':443',
                          null, null);

//rpc Get(BillingsRequest) returns (BillingDataList);
const GetBilling = ( {data, result, error=null, paths=[], metadata={}} ) => {

  const billing = new BillingData()
  Object.entries(data).forEach( ([key, value]) => {
    billing["set"+key.capitalize()](value)
  })

  const fields = new FieldMask();
  fields.setPathsList( paths );

  const billings = new BillingDataList()
  billings.setBillingsList(
    [ billing ]
  )

  const request = new BillingsRequest();
  request.setBillingsdata( billings );
  request.setFields( fields );

  const call = BillingService.get(
    request, metadata, 
    (err, response) => {
      if (err) {
        console.log(`Unexpected error for GetBilling: code = ${err.code}, message = "${err.message}"`);
        if (error){
          error({
            msg: err.message,
            type: 'ErrorMessage'
          })
        }
        result()
      } else {
        console.log("GetBilling", response.toObject().billingsList );
        if (error){
          error()
        }
        result(
          response.toObject().billingsList
        )
      }
  });
}

export { 
  GetBilling
};

File: src/backend/orchestrator/src/service/billing_client.py
import os
import grpc

import proto.query.billing.query_billing_pb2 as query_billing_pb2
import proto.query.billing.query_billing_pb2_grpc as query_billing_pb2_grpc


from google.protobuf import field_mask_pb2

billing_address=os.environ['billing-address']
billing_port=os.environ['billing-port']

channel_string = f"{billing_address}:{billing_port}"

class BillingClient:

  def __BillingData(datas):
    print(f'__BillingData datas:{datas}', flush=True)
    return [ query_billing_pb2.BillingData( **data ) for data in datas if data ]
  
  def __init__(self):
    # instantiate a channel
    self.channel  = grpc.insecure_channel(channel_string)
    # bind the client and the server
    self.stub = query_billing_pb2_grpc.BillingStub(self.channel)

  def get_billing(self, datas, paths):
    print(f'get_billing 1', flush=True)
    billings = BillingClient.__BillingData(datas=datas)

    print(f'get_billing  2', flush=True)
    response = self.stub.Get(
              query_billing_pb2.BillingsRequest(
                billingsData=query_billing_pb2.BillingDataList(
                    billings=billings
                  ),
                fields=field_mask_pb2.FieldMask( 
                  paths=paths 
                )
              )
            )
    print(f'get_billing 3', flush=True)
    return response

